src/log/log.cr

```diff
@@ -0,0 +1,37 @@
+class Log
+  getter source : String
+  getter backend : Backend?
+  getter level : Severity
+
+  # :nodoc:
+  def initialize(@source : String, @backend : Backend?, @level : Severity)
+  end
+
+  # Change this log severity level filter
+  def level=(value : Severity)
+    @level = value
+    if (backend = @backend).responds_to?(:level=)
+      backend.level = value
+    end
+    value
+  end
+
+  {% for method, severity in {
+                               debug:   Severity::Debug,
+                               verbose: Severity::Verbose,
+                               info:    Severity::Info,
+                               warn:    Severity::Warning,
+                               error:   Severity::Error,
+                               fatal:   Severity::Fatal,
+                             } %}
+
+    # Logs a message if the logger's current severity is lower or equal to `{{severity}}`.
+    def {{method.id}}(*, exception : Exception? = nil)
```

Comment:
It will prevent using a macro :+1:

Reply:
Why would you need a parametric severity when emitting a log?

That will also prevent doing some optimizations like removing in compile time all `.debug` methods for example.

Reply:
> That will also prevent doing some optimizations like removing in compile time all `.debug` methods for example.

Those optimizations don't need to be present when using parametric severity.

We can wait for someone to complain, if you want, though.

Reply:
Every log library or framework that I used, it's `log.info`, `log.debug`. I see no reason why we need to write `Log.log(:info) { "Hello" }` instead of `Log.info { "hello" }`.

Reply:
And then the log level you use is never parameterized in each invocation.

Reply:
I think it'd be nice to have a generic `log` , but having `debug/etc` is a must (IMO). The log that accepts severity is nice when you have a dynamic value. For example you have an ENV, or a value stored in Redis/Memory that is used to determine the level. Another example is determining the log level based on something else that can't be determined at compile time

```crystal
def run_query
  log_level = if query.was_slow?
    :info
  else
     :debug
   end
   Log.write(log_level, query.to_sql)
end
```

Definitely helpful to have if needed, but the concrete methods should be preferred in most cases IMO

Reply:
Hmm, I thought @RX14 was talking about `abstract def log` in addition to the macro magic, **not** meant to replace it altogether.